Add battery_charging & last_seen attributes to GPSLogger

pnbruckner opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
The GPSLogger app can provide battery charging status and an update timestamp. Allow configuring these in the Android app and, if present in the received webhook packet, add them as attributes to the GPSLogger tracker entities.

Note that it is possible for packets from the Android app to be received out of order due to network delays, etc. If this happens, the `device_tracker` entity will move "backwards" (because it will be updated with old GPS data.) This can be prevented by utilizing the `last_seen` timestamp. If a packet is received with a `last_seen` value that is _before_ the previously received packet's `last_seen` value, skip the newly received packet (and log a debug message to document this has happened.) This change implements that safeguard (which the Google Maps integration also implements, and the Life360 integration did before it was removed.)

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: TBD

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

pnbruckner committed:
Allow battery_charging & last_seen attributes

Skip update if last_seen went backwards.

pnbruckner committed:
Warn if app not configured to send last_seen

pnbruckner committed:
Fix gpslogger test_load_unload_entry

pnbruckner committed:
Add tests for last_seen attribute

MartinHjelmare commented on homeassistant/components/gpslogger/device_tracker.py:
```diff
@@ -137,40 +148,53 @@ async def async_added_to_hass(self) -> None:
-            ATTR_ALTITUDE: attr.get(ATTR_ALTITUDE),
             ATTR_ACTIVITY: attr.get(ATTR_ACTIVITY),
+            ATTR_ALTITUDE: attr.get(ATTR_ALTITUDE),
+            ATTR_BATTERY_CHARGING: attr.get(ATTR_BATTERY_CHARGING),
```
Battery charging should be a binary sensor entity.

pnbruckner replied:
I don't care so much about this piece of data, but why make this special when all other miscellaneous data are attributes?

MartinHjelmare replied:
It's our current standard to put all measurements that are relevant on their own as separate sensor entities.

pnbruckner replied:
So, does that mean I need to pull out all the others, too, and make them separate sensor entities? At a minimum, shouldn't that also apply to activity, altitude, battery_level, direction & speed? And can I assume, since you did not reply to my comments about last_seen, it's ok to leave it as an attribute on the device_tracker entity?

MartinHjelmare replied:
Existing attributes don't need to be moved at this time, since that would be a breaking change.

We don't allow new additions of attributes that should be sensors.

Last seen is not allowed as an attribute.

pnbruckner replied:
Then there's no point.

MartinHjelmare commented on homeassistant/components/gpslogger/device_tracker.py:
```diff
@@ -137,40 +148,53 @@ async def async_added_to_hass(self) -> None:
+            ATTR_ALTITUDE: attr.get(ATTR_ALTITUDE),
+            ATTR_BATTERY_CHARGING: attr.get(ATTR_BATTERY_CHARGING),
             ATTR_DIRECTION: attr.get(ATTR_DIRECTION),
+            ATTR_LAST_SEEN: last_seen,
```
Last seen should be a timestamp sensor entity.

pnbruckner replied:
It is much easier to compare multiple device_tracker entities for a single physical device (e.g., when using multiple presence detection integrations) when last_seen is an attribute, not a separate entity. And there is already precendence for a timestamp like this to be an attribute. E.g., last_seen in (the now defunct) Life360, Google Maps' last_seen, Tile's last_timestamp, and possibly even asuswrt's last_time_reachable. It is crucial for my custom composite integration that many, many people use.

pnbruckner replied:
FWIW, I understand splitting out miscellaneous data into separate sensors as opposed to the original tendency to keep adding attributes to an existing entity. I have no problem with that in general.

However, in this case, I would argue that this is a perfect example of what an attribute should be used for. It is adding information about the main state (latitude, longitude & accuracy) that makes using that data more accurate. I.e., it describes when the data actually changed (as opposed to the last_updated field, that just indicates when the data was received by HA, which can be, and often is, very different.)

pnbruckner committed:
Convert last_seen back to datetime after restore

Accuracy should be 0 instead of None.

Handle case when restored without last_seen attribute.

pnbruckner committed:
Add more tests for full coverage
